src/AspNet.Security.OAuth.Basecamp/BasecampAuthenticationHandler.cs

```diff
@@ -0,0 +1,71 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AspNet.Security.OAuth.Basecamp
+{
+    public class BasecampAuthenticationHandler : OAuthHandler<BasecampAuthenticationOptions>
+    {
+        public BasecampAuthenticationHandler(
+            [NotNull] IOptionsMonitor<BasecampAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(
+            [NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] OAuthTokenResponse tokens)
+        {
+            using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred while retrieving the user profile.");
+            }
+
+            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+
+            if (payload.RootElement.TryGetProperty("identity", out var identityProperty))
```

Comment:
What requires the need to test for this property here and fork the logic?

Reply:
I noticed the FoursquareAuthenticationHandler.cs doing this pattern too.  It seemed wise to add the same check for the BasecampAuthenticationHandler.cs.  Basecamp could change their "identity" object, and if that were to happen, I wanted the code to handle it gracefully.  But I would be happy to remove the check, if you prefer it not be there.

Reply:
What are your thoughts?

Reply:
Looking at the Foursquare provider's history, it looks like it always had a lot of null checking as it went down the JSON object's hierarchy out of an abundance of caution.

I don't think we need that here. Instead, we can use the `ClaimActions.MapJsonSubKey(string, string, string)` method added in .NET Core 3.0 in the `BasecampAuthenticationOptions` class, and then this can just be `context.RunClaimActions()`. I've marked up the changes in the diff.

Reply:
Great suggestion, I have made the suggested code changes.  They are part of this pull-request.  Thanks